src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -315,6 +321,51 @@ public WebSession authenticate(SessionManagementMethod sessionManagementMethod,
 			// Return the web session as extracted by the session management method
 			return sessionManagementMethod.extractWebSession(msg);
 		}
+		
+		/**
+		 * Modifies the input {@code requestMessage} by replacing fresh anti-CSRF token value in the request body.  
+		 * It first checks if the input {@code requestMessage} has an anti-CSRF token.
+		 * If yes, then it makes a {@code GET} request to the login page in order to renew the anti-CSRF token.
+		 * Then it modifies the input {@code requestMessage} with the regenerated anti-CSRF token value.
+		 * If the first {@code POST} login request does not have anti-CSRF token
+		 * then the input {@code requestMessage} is left as it is.
+		 * 
+		 * @param	requestMessage	the login request message with correct credential
+		 * @param	user 			the user of the context
+		 * @throws	IOException		if problem occurs when sending login request to get login page.
+		 */
+		private void replaceAntiCsrfTokenValueIfRequired(HttpMessage requestMessage, User user) throws IOException {
+			if(extAntiCsrf == null) {
+				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
+			}
+			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+  			if(extAntiCsrf != null) {
+				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
+			}
+			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
+			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
+				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+			}
+			
+			if (antiCsrfTokenOfFirstLoginMsg != null) {
+				HttpMessage loginMsgToRegenerateAntiCsrfToken = antiCsrfTokenOfFirstLoginMsg.getMsg().cloneAll();
+				//to make sure cookies from the loginMsgToRegenerateAntiCsrfToken response
```

Comment:
I think this comment could be removed the docs of `setRequestingUser` already imply that.